### PR TITLE
feat(admin) sugar method to delete upstream target

### PR DESCRIPTION
**Pull request by p0pr0ck5**

### CODE REVIEW

File: spec/02-integration/03-admin_api/09-targets_routes_spec.lua [hunk 1]
```diff
@@ -320,4 +320,52 @@ describe("Admin API", function()
       end)
     end)
   end)
+
+  describe("/upstreams/{upstream}/targets/{target}", function()
+    describe("DELETE", function()
+      local target
+
+      before_each(function()
+        assert(helpers.dao.targets:insert {
```
Comment (Tieske) [inline spec/02-integration/03-admin_api/09-targets_routes_spec.lua]: shouldn't an upstream be created here? reusing an existing one might interfere with the test results here.

File: kong/api/routes/upstreams.lua [hunk 2]
```diff
@@ -160,4 +103,25 @@ return {
...
+    before = function(self, dao_factory, helpers)
+      crud.find_upstream_by_name_or_id(self, dao_factory, helpers)
+      self.params.upstream_id = self.upstream.id
+    end,
+
+    DELETE = function(self, dao_factory)
+      balancer.clean_history(self.params.upstream_id, dao_factory)
+
+      -- this is just a wrapper around POSTing a new target with weight=0
+      local target  = self.params
```
Comment (thibaultcha) [inline kong/api/routes/upstreams.lua]: It would be more explicit to use `self.upstream` here

File: kong/api/routes/upstreams.lua [hunk 3]
```diff
@@ -160,4 +103,25 @@ return {
...
+      target.weight = 0
+
+      local data, err = dao_factory.targets:insert(target)
+      if err then
+        return app_helpers.yield_error(err)
+      end
+
+      return responses.send_HTTP_NO_CONTENT()
```
Comment (thibaultcha) [inline kong/api/routes/upstreams.lua]: Wondering if we should not send `HTTP 201` still here. Yes, this is a DELETE and stays consistent with the rest of the API, but it isn't a traditional DELETE endpoint. It is known that this is a sugar method. Might bring confusion, but might also raise awareness about targets. However, this is mostly internal implementation details, so probably good to hide it away.

Comment (p0pr0ck5) [inline kong/api/routes/upstreams.lua]: I think if we were to return 201 here, we might as well not bother with this method at all ;)

tbh, i dont even think this is a valid PR. I don't think adding this endpoint makes it any easier for consumers to work with the upstream/target interface. But that's just my $0.02 ;)

File: spec/02-integration/03-admin_api/09-targets_routes_spec.lua (hunk 1 above)
Comment (p0pr0ck5) [inline spec/02-integration/03-admin_api/09-targets_routes_spec.lua]: But it doesn't :)

My understanding of `before_each` means that it runs before each nested test, meaning that we recreate the upstream and shut down kong after every test. And actually, the anti-pattern that I used in the previous PR to create `upstream3` is unnecessary as well.

Nevertheless, I'll add it.

File: kong/api/routes/upstreams.lua (hunk 2 above)
Comment (p0pr0ck5) [inline kong/api/routes/upstreams.lua]: Hmm? `self.params` represents the target (a table with a  `target` key), not an upstream.

File: kong/api/routes/upstreams.lua [hunk 4]
```diff
@@ -46,65 +47,7 @@ return {
...
-            -- in case another kong-node does the same cleanup simultaneously
-            cnt = cnt + 1 
-          end
-
-          ngx.log(ngx.INFO, "[admin api] Finished cleanup of target table",
-            " for upstream ", tostring(self.params.upstream_id),
-            " removed ", tostring(cnt), " target entries")
-        end
-      end
+      balancer.clean_history(self.params.upstream_id, dao_factory)
```
Comment (Tieske) [inline kong/api/routes/upstreams.lua]: I'm having doubts on whether we should be moving this code. So far (at least the way I see it) the code for the management api and the Kong runtime have remained fairly separate. By moving this code into the balancer module, we change that, as this code is strictly management api.

Not sure though...

@thibaultcha wat do you think?

File: kong/api/routes/upstreams.lua (hunk 2 above)
Comment (Tieske) [inline kong/api/routes/upstreams.lua]: `self.params` represents the request parameters, with modifications from the `before` handler.

I'd consider it cleaner to create a new target table here

```lua
local data, err = dao_factory.targets:insert( {
  target = params.target,
  upstream_id = self.upstream.id,
  weight = 0,
})
```
Then line 109 above can be dropped as well.

File: kong/api/routes/upstreams.lua (hunk 4 above)
Comment (p0pr0ck5) [inline kong/api/routes/upstreams.lua]: Wasn't sure about the style of putting local functions in `routes` files, only reason I thought to put it in `kong/core/balancer`. Ill move it back :)

File: kong/api/routes/upstreams.lua (hunk 2 above)
Comment (p0pr0ck5) [inline kong/api/routes/upstreams.lua]: wfm!

**Comment by Tieske** [comment]
As this is fully backward compatible, this could as well go in `master` imo, so it can be shipped with `0.10.1`

**Comment by p0pr0ck5** [comment]
@Tieske that also brings in https://github.com/Mashape/kong/pull/2230, since this was based on that PR. if youre okay with that, ill merge these to master and back-merge to next

**Comment by Tieske** [comment]
tagged them both for milestone `0.10.1`

**Pull request merged**